### DISCUSSION TITLE
[Widget] Markdown Preview

### DISCUSSION THREAD

**Discussion by zerebos** [TriliumNext/Trilium#4172] [Show and tell]
# Trilium-MarkdownPreview - [Link](https://github.com/rauenzi/Trilium-MarkdownPreview/)

A widget for Trilium Notes for live previewing markdown files with support for anchors, images, and sync scroll.

## Why?
I know Trilium isn't a markdown editor, but I still use markdown for other things like my GitHub READMEs and I like storing everything in one place--Trilium. The only problem was, I couldn't preview my markdown files, and that's super helpful to have before I go pushing a commit. After looking, I couldn't find an existing widget for it, so I decided to make my own.

## Preview
![Banner](https://user-images.githubusercontent.com/6865942/260894889-93194c61-eee3-49fd-8ed8-970f1476539c.png)

## Features
- Preview markdown notes in real time
- Global and per-note styles
- Embed local images
- Clickable in-page anchor links
- Syncable scrollbars
- GFM-compliant with tables and more

## Installation

1. Create a new code note with `JS Frontend` type. Copy and paste [widget.js](https://github.com/rauenzi/Trilium-MarkdownPreview/blob/main/src/widget.js) into the note.
    - Be sure to add the `#widget` attribute to the note
1. Attach the [marked.min.js](https://github.com/rauenzi/Trilium-MarkdownPreview/blob/main/lib/marked.min.js) file to the note either via drag-n-drop or import.
1. (optional) Create one (or more) child code notes of `CSS` type and fill it with any global markdown styles.

### Syntax Highlighting (optional)
If you want your codeblocks to have syntax highlighting, follow these steps:

1. Attach the [highlight.min.js](https://github.com/rauenzi/Trilium-MarkdownPreview/blob/main/lib/highlight.min.js) file to the note either via drag-n-drop or import.
1. (optional) Attach the pre-made syntax styles from [styles.css](https://github.com/rauenzi/Trilium-MarkdownPreview/blob/main/src/styles.css) as a child note of the [main widget](#installation).

## Usage

Create a code note with markdown type. Then, to make a note previewable, simply add the `#markdownPreview` attribute to it. (You may have to switch notes afterward for the preview to appear.)

After that, just start writing in markdown on the code editor side. You'll see the preview update with you as you go.

## More Info
For more information on usage configuration, or to see some **demo videos**, go checkout [the repository on GitHub](https://github.com/rauenzi/Trilium-MarkdownPreview/).

**Comment by Nriver** [comment]
It's a legendary enhancement for Markdown support.

**Comment by Nriver** [comment]
The code highlight works great!

![ksnip_20230818-094223](https://github.com/zadam/trilium/assets/6752679/908ed55b-0c6e-439f-8ecc-9f0b91d9ccfc)

**Reply by zerebos** [reply]
Glad to hear it 😄

**Comment by zerebos** [comment]
Dropping this here for anyone that may be interested in markdown-related stuff: I'm working on a new plugin/widget...

https://github.com/zadam/trilium/assets/6865942/34e3533b-fe92-4918-a6bd-d66a357fa25c

**Reply by Nriver** [reply]
Feel free to add your creations to the list https://github.com/Nriver/awesome-trilium

**Reply by zerebos** [reply]
I definitely will when it's done! I just wanted to show this off as a proof-of-concept! :smile:

**Comment by yqook** [comment]
The widget is amazing! It makes Trilium has more possibility. But there is a bug:
If you have a direct child note type of `CSS` , the `'CSS'` will be applied globally when you view the parent note.

**Reply by zerebos** [reply]
I'm not sure I understand, can you be a bit more specific? The `CSS` note is a child of the widget? Or of a markdown note?

If it's the child of a markdown note, do you mean it's also affecting _other_ markdown notes?

**Reply by yqook** [reply]
Sorry for the poor english. Hope the picture can explain the problem clearly:
![image](https://github.com/zadam/trilium/assets/30325775/f359d17a-d35b-4a05-bb50-32d81816ad35)

**Reply by yqook** [reply]
Add:
- The parent note "Any type of parent note" doesn't have any attribute.

**Reply by zerebos** [reply]
Nriver ended up making a similar issue on the repo which helped me debug it. Change line `155` from
```js
        if (this.note) {
```
to
```js
        if (this.note && this.isEnabled()) {
```

And then reload and the bug should be fixed!
